File: backend/app/schemas/research.py
from typing import List, Optional, Dict, Any
from enum import Enum
from pydantic import BaseModel, Field
from app.schemas.paper import PaperResponse
from app.schemas.claim import ClaimResponse
from app.schemas.trust import TrustAnalysis, ConfidenceLevel

class WorkflowStage(str, Enum):
    IDLE = "IDLE"
    PLANNING = "PLANNING"
    RETRIEVAL = "RETRIEVAL"
    RANKING = "RANKING"
    EVIDENCE_EXTRACTION = "EVIDENCE_EXTRACTION"
    CLAIM_VERIFICATION = "CLAIM_VERIFICATION"
    CONTRADICTION_DETECTION = "CONTRADICTION_DETECTION"
    TRUST_ANALYSIS = "TRUST_ANALYSIS"
    RED_TEAM = "RED_TEAM"
    RESEARCH_GAPS = "RESEARCH_GAPS"
    JUDGE = "JUDGE"
    SYNTHESIS = "SYNTHESIS"
    FIREWALL = "FIREWALL"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"

class ResearchMode(str, Enum):
    QUICK_RESEARCH = "QUICK_RESEARCH"
    DEEP_RESEARCH = "DEEP_RESEARCH"
    LITERATURE_REVIEW = "LITERATURE_REVIEW"
    PAPER_RANKING = "PAPER_RANKING"
    PAPER_AUDIT = "PAPER_AUDIT"
    REPLICATION_ANALYSIS = "REPLICATION_ANALYSIS"
    RESEARCH_COMPARISON = "RESEARCH_COMPARISON"
    RESEARCH_GAP_FINDER = "RESEARCH_GAP_FINDER"

class ResearchRequest(BaseModel):
    query: str = Field(..., description="Research question to investigate", min_length=5)
    max_papers: int = Field(default=10, ge=1, le=50, description="Max academic papers to retrieve")
    mode: ResearchMode = Field(default=ResearchMode.DEEP_RESEARCH, description="Selected research mode")

class ResearchPlan(BaseModel):
    objective: str
    subquestions: List[str] = Field(default_factory=list)
    search_queries: List[str] = Field(default_factory=list)
    required_evidence: List[str] = Field(default_factory=list)
    ambiguities: List[str] = Field(default_factory=list)

class Contradiction(BaseModel):
    contradiction_id: str
    claim_a_id: str
    claim_b_id: str
    paper_a_id: str
    paper_b_id: str
    type: str = Field(..., description="direct_contradiction, partial_contradiction, different_datasets, different_metrics, different_populations, different_conditions")
    description: str

class RedTeamFinding(BaseModel):
    category: str = Field(..., description="weak_methodology, unsupported_assumptions, overgeneralization, dataset_limitations, citation_problems, reproducibility_concerns")
    challenge: str
    severity: str = Field(default="MEDIUM", description="HIGH, MEDIUM, LOW")
    target_claim_id: Optional[str] = None

class ResearchGap(BaseModel):
    gap_id: str
    title: str
    description: str
    understudied_area: str
    connected_paper_ids: List[str] = Field(default_factory=list)

class FinalJudgement(BaseModel):
    status: ConfidenceLevel
    main_conclusion: str
    reasoning: str
    firewall_passed: bool = Field(default=True)
    firewall_revisions: List[str] = Field(default_factory=list)

class ResearchStatusResponse(BaseModel):
    session_id: str
    query: str
    current_stage: WorkflowStage
    progress_percentage: int = Field(default=0, ge=0, le=100)
    papers_found_count: int = 0
    relevant_papers_count: int = 0
    claims_extracted_count: int = 0
    supported_claims_count: int = 0
    contradicted_claims_count: int = 0
    insufficient_claims_count: int = 0
    plan: Optional[ResearchPlan] = None
    papers: List[PaperResponse] = Field(default_factory=list)
    claims: List[ClaimResponse] = Field(default_factory=list)
    contradictions: List[Contradiction] = Field(default_factory=list)
    trust_analysis: Optional[TrustAnalysis] = None
    red_team_findings: List[RedTeamFinding] = Field(default_factory=list)
    research_gaps: List[ResearchGap] = Field(default_factory=list)
    final_judgement: Optional[FinalJudgement] = None
    report_markdown: Optional[str] = None
    error: Optional[str] = None

File: frontend/types/research.ts
export type VerificationStatus = 
  | "SUPPORTED"
  | "PARTIALLY_SUPPORTED"
  | "CONTRADICTED"
  | "INSUFFICIENT_EVIDENCE";

export type ConfidenceLevel = 
  | "HIGH_CONFIDENCE"
  | "MODERATE_CONFIDENCE"
  | "UNCERTAIN"
  | "INSUFFICIENT_EVIDENCE";

export type WorkflowStage = 
  | "IDLE"
  | "PLANNING"
  | "RETRIEVAL"
  | "RANKING"
  | "EVIDENCE_EXTRACTION"
  | "CLAIM_VERIFICATION"
  | "CONTRADICTION_DETECTION"
  | "TRUST_ANALYSIS"
  | "RED_TEAM"
  | "RESEARCH_GAPS"
  | "JUDGE"
  | "SYNTHESIS"
  | "FIREWALL"
  | "COMPLETED"
  | "FAILED";

export type ResearchMode = 
  | "QUICK_RESEARCH"
  | "DEEP_RESEARCH"
  | "LITERATURE_REVIEW"
  | "PAPER_RANKING"
  | "PAPER_AUDIT"
  | "REPLICATION_ANALYSIS"
  | "RESEARCH_COMPARISON"
  | "RESEARCH_GAP_FINDER";

export interface ClarificationQuestion {
  id: string;
  question: string;
  options: string[];
  suggested_default: string;
}

export interface ClarificationResponse {
  query: string;
  followup_questions: ClarificationQuestion[];
  suggested_refinements: string[];
}

export interface PaperRankBreakdown {
  relevance_score: number;
  citation_score: number;
  methodology_score: number;
  reproducibility_score: number;
  provenance_score: number;
  recency_score: number;
  evidence_quality_score: number;
  overall_rank_score: number;
  ranking_explanation: string;
}

export interface PaperAuditResult {
  paper_id: string;
  traceability_chain: string[];
  mismatches_found: string[];
  audit_passed: boolean;
}

export interface ReplicationAnalysisResult {
  paper_id: string;
  main_experiment: string;
  required_datasets: string[];
  methods: string[];
  metrics: string[];
  implementation_requirements: string[];
  reproducibility_score: number;
  sufficient_info_available: boolean;
  missing_information: string[];
}

export interface Paper {
  id: string;
  paper_id: string;
  title: string;
  authors: string[];
  year?: number;
  abstract?: string;
  doi?: string;
  url?: string;
  source: string;
  citation_count: number;
  reproducibility_score?: number;
  bias_score?: number;
  rank_breakdown?: PaperRankBreakdown;
  audit_result?: PaperAuditResult;
  replication_analysis?: ReplicationAnalysisResult;
}

export interface Claim {
  id: string;
  claim_id: string;
  statement: string;
  paper_id: string;
  paper_title?: string;
  evidence_text: string;
  methodology?: string;
  dataset?: string;
  sample_size?: string;
  metrics?: string;
  results?: string;
  limitations?: string;
  verification_status: VerificationStatus;
  confidence_score: number;
  explanation: string;
  contradicting_claim_ids?: string[];
}

export interface Contradiction {
  contradiction_id: string;
  claim_a_id: string;
  claim_b_id: string;
  paper_a_id: string;
  paper_b_id: string;
  type: string;
  description: string;
}

export interface TrustScoreBreakdown {
  evidence_quality: number;
  source_reliability: number;
  independent_agreement: number;
  methodology_quality: number;
  reproducibility: number;
  citation_support: number;
}

export interface TrustAnalysis {
  overall_trust_score: number;
  confidence_level: ConfidenceLevel;
  breakdown: TrustScoreBreakdown;
  explanation: string;
  knowns: string[];
  unknowns: string[];
  missing_evidence: string[];
}

export interface RedTeamFinding {
  category: string;
  challenge: string;
  severity: "HIGH" | "MEDIUM" | "LOW";
  target_claim_id?: string;
}

export interface ResearchGap {
  gap_id: string;
  title: string;
  description: string;
  understudied_area: string;
  connected_paper_ids: string[];
}

export interface FinalJudgement {
  status: ConfidenceLevel;
  main_conclusion: string;
  reasoning: string;
  firewall_passed: boolean;
  firewall_revisions?: string[];
}

export interface ResearchStatus {
  session_id: string;
  query: string;
  current_stage: WorkflowStage;
  progress_percentage: number;
  papers_found_count: number;
  relevant_papers_count: number;
  claims_extracted_count: number;
  supported_claims_count: number;
  contradicted_claims_count: number;
  insufficient_claims_count: number;
  plan?: {
    objective: string;
    subquestions: string[];
    search_queries: string[];
    required_evidence: string[];
    ambiguities: string[];
    breadth?: number;
    depth?: number;
  };
  papers: Paper[];
  claims: Claim[];
  contradictions: Contradiction[];
  trust_analysis?: TrustAnalysis;
  red_team_findings: RedTeamFinding[];
  research_gaps: ResearchGap[];
  final_judgement?: FinalJudgement;
  report_markdown?: string;
  error?: string;
}

export interface VisualConcept {
  concept_id: string;
  name: string;
  concept_type: "ENTITY" | "VARIABLE" | "METRIC" | "DATASET" | "METHOD";
  description: string;
  evidence_snippet: string;
  paper_id?: string;
}

export interface VisualOpportunity {
  opportunity_id: string;
  visual_type: "METHODOLOGY_FLOW" | "RESEARCH_TIMELINE" | "EVIDENCE_NETWORK" | "CONCEPT_MATRIX";
  confidence: number;
  justification: string;
}

export interface TimelineMilestone {
  milestone_id: string;
  year_or_date: string;
  title: string;
  description: string;
  paper_id?: string;
  doi?: string;
}

export interface DiagramNode {
  node_id: string;
  label: string;
  node_type: "INPUT" | "PROCESS" | "MODEL" | "OUTPUT" | "CLAIM" | "EVIDENCE";
  description: string;
  paper_id?: string;
  evidence_text?: string;
}

export interface DiagramEdge {
  edge_id: string;
  source_id: string;
  target_id: string;
  label: string;
  relationship_type: "FEEDS_INTO" | "PRODUCES" | "SUPPORTS" | "CONTRADICTS";
}

export interface AnimationKeyframe {
  frame_index: number;
  step_title: string;
  description: string;
  active_node_ids: string[];
  camera_focus: string;
  duration_ms: number;
}

export interface AnimationSchema {
  fps: number;
  total_duration_ms: number;
  keyframes: AnimationKeyframe[];
}

export interface VisualResearchPlan {
  session_id: string;
  paper_id: string;
  paper_title: string;
  concepts: VisualConcept[];
  opportunities: VisualOpportunity[];
  timeline: TimelineMilestone[];
  methodology_nodes: DiagramNode[];
  methodology_edges: DiagramEdge[];
  evidence_nodes: DiagramNode[];
  evidence_edges: DiagramEdge[];
  animation_schema: AnimationSchema;
  validation_passed: boolean;
  validation_notes: string[];
}
